### PR TITLE
[WIP] Initial Handlebars/Glimmer support

### PR DESCRIPTION
**Pull request by azz**
Experimenting with HTML/Handlebars/Mustache/Glimmer support for #1882

Picked the glimmer parser as I thought it would have the widest range of support.

I've ran into a few issues so far:

* Some handlebars features (`{{> foo}}`, `{{../foo}}`) are not supported. [source](https://github.com/glimmerjs/glimmer-vm/blob/master/packages/%40glimmer/syntax/lib/parser/handlebars-node-visitors.ts#L158)
* `<!doctype>` doesn't parse.

The formatting is also super ugly at the moment because I figured the best way to go would be re-use some of the JSX formatting logic.

I also moved the postcss printer to a separate file.

### CODE REVIEW

File: src/options.js [hunk 1]
```diff
@@ -33,6 +33,8 @@ function normalize(options) {
     normalized.parser = "postcss";
   } else if (/\.(ts|tsx)$/.test(filepath)) {
     normalized.parser = "typescript";
+  } else if (/\.(html|hbs|vue)$/.test(filepath)) {
```
Comment (SimenB) [inline src/options.js]: `.moustache` as well

Comment (azz) [inline src/options.js]: Or `.mustache`?

Comment (azz) [inline src/options.js]: I'll go with `/mo?ustache/` :p

Comment (SimenB) [inline src/options.js]: hah, indeed!

**Comment by azz** [comment]
Not sure what's going on with the build on Node.js v4.

```
FAIL  tests/html_glimmer/jsfmt.spec.js
  ● Test suite failed to run
    SyntaxError: Unexpected token =
      
      at ScriptTransformer._transformAndBuildScript (node_modules/jest-runtime/build/ScriptTransformer.js:289:17)
      at Object.<anonymous> (node_modules/@glimmer/syntax/dist/commonjs/index.js:4:16)
      at parse (src/parser-glimmer.js:6:25)
```
`src/parser-glimmer.js:6:25` is:
```ts
  const glimmerSyntax = require("@glimmer/syntax");
```

**Comment by azz** [comment]
Misleading error. Tracked it down to glimmer's source, turns out they're building JS without node v4 support:

`node_modules/@glimmer/syntax/dist/commonjs/lib/parser.js:33`:
```
     constructor(source, options = {}) {
```

**Comment by vjeux** [comment]
Omg, this is so exciting!

**Comment by vjeux** [comment]
Cc @mixonic, @wycats

**Comment by azz** [comment]
Not having much luck with patching `@glimmer/syntax`, will need to clone it and do a custom build tomorrow.

**Comment by azz** [comment]
@vjeux I disabled the new tests in case you want to merge this.

**Comment by vjeux** [comment]
Is glimmer parser a fully compliant html5 parser?

**Comment by vjeux** [comment]
Looks like it's not using an html5 compliant parser and doesn't intend to be: https://github.com/tildeio/simple-html-tokenizer


That makes me uneasy to include inside of prettier as people are likely to throw all sorts of crazy code at it.

**Comment by wycats** [comment]
simple-html-tokenizer is intended to be a compliant tokenizer for HTML that could be found inside of an HTML body, not including scripts. It doesn't always error-correct (in cases where the HTML spec specifies that a particular character is an error, we sometimes choose to throw an exception rather than follow the error correction behavior). 

We think this is a good trade-off of size and simplicity to compliance for the use case of HTML templates. We also have found that Glimmer and Ember users really appreciate early errors for invalid HTML, which can be confusing when read later on.

**Comment by wycats** [comment]
Is your concern about applying prettier to Handlebars templates that are emitting HTML but less strict than the Glimmer parser?

**Comment by vjeux** [comment]
I'd love for prettier to support regular html, vue files, angular ones, glimmer templates, mustache, handlebar... I don't have a lot of context around the different ways each of them parse code, but if we can find a single parser that is able to parse them in a way that if we print them, they have the same semantics, that would be awesome. I'm not sure if one exists though.

**Comment by vjeux** [comment]
It looks like glimmer parser is simple-html-tokenizer for the html parsing and then uses handlebars itself to compile the templates: `import { parse } from "./handlebars/compiler/base";`.

Would it be possible to use a real html5 parser like https://github.com/inikulin/parse5 which is used by angular2 and polymer to parse the html and then use handlebars parser for the templates part?

**Comment by vjeux** [comment]
Vue also uses a small handrolled html parser: https://github.com/vuejs/vue/blob/dev/src/compiler/parser/html-parser.js

**Comment by azz** [comment]
Looking into parse5, it supports two AST formats:

>default: TreeAdapter
>Default tree format for parse5.

> htmlparser2: TreeAdapter
> Quite popular htmlparser2 tree format (e.g. used by cheerio and jsdom).

Any preferences?

**Comment by azz** [comment]
Angular also appears to use a handmade parser.

https://github.com/angular/angular/blob/master/packages/compiler/src/ml_parser/html_parser.ts

**Comment by wycats** [comment]
@vjeux not quite. Stuff like this needs to be parsed with an integrated parser:

```hbs
<div title="{{t "frontpage.title"}}">
```

Otherwise, the HTML parser will think that the second quotation mark (after `t`) is closing the title attribute. What we do instead is parse as handlebars first, then tokenize the leftover content using the HTML tokenizer library above.

**Comment by vjeux** [comment]
@wycats do you know if is glimmer parser able to parse all the angular 2 and vue files as well?

**Comment by aboyton** [comment]
I know it's a lot to ask, but any idea if glimmer can also parse underscore/lodash templates?

**Comment by vjeux** [comment]
While we figure out which parser to use, would you mind shipping the part that splits css in its own file? This is good and should help the folks trying to integrate graphql: https://twitter.com/stubailo/status/871584723524505600

**Comment by azz** [comment]
Done. #1969

**Comment by vjeux** [comment]
Are you going to keep working on this PR or should you open a new one?

**Comment by azz** [comment]
I'll take a look at this today.

**Comment by azz** [comment]
Gah, didn't get a chance to look further into this today. Got #2108 merged though.

Since we've recently merged a separate HTML parser, I'm thinking it might be wise to stabilise that before merging in another parser.

**Comment by vjeux** [comment]
I'm going to close it. Feel free to re-open it when you'll be working on it again.

**Pull request closed without merging**